heartbeat/iface-macvlan

```diff
@@ -0,0 +1,363 @@
+        Define the MACVLAN NAME.
+      </shortdesc>
+      <content type="string"/>
+    </parameter>
+
+    <parameter name="mode" unique="1">
+      <longdesc lang="en">
+        Define the name of the MACVLAN mode (currently only bridge is supported).
+      </longdesc>
+      <shortdesc lang="en">
+        Mode of the macvlan.
+      </shortdesc>
+      <content type="string" default="${OCF_RESKEY_mode_default}" />
+    </parameter>
+
+    <parameter name="mac">
+     <longdesc lang="en">
+      Set the interface MAC address explicitly.
+     </longdesc>
+     <shortdesc lang="en">MAC address of the macvlan</shortdesc>
+     <content type="string" />
+</parameter>
+  </parameters>
+
+  <actions>
+    <action name="start"        timeout="30s" />
+    <action name="stop"         timeout="20s" />
+    <action name="status"       timeout="20s" depth="0" interval="10s" />
+    <action name="monitor"      timeout="20s" depth="0" interval="10s" />
+    <action name="meta-data"    timeout="5s" />
+    <action name="validate-all" timeout="20s" />
+  </actions>
+</resource-agent>
+END
+}
+
+# check if the interface is admin up/down
+
+iface_is_up() {
+	if ! $IP2UTIL -o link show $1 | \
+	    sed -e 's#.*<##g' -e 's#>.*##' -e 's#LOWER_UP##g' | \
+	    grep -q UP; then
+		return 1
+	fi
+	return 0
+}
+
+# check if the slaves have link layer up/down
+# see kernel network documentation on meaning of LOWER_UP flag
+# for more in depth explanation on how it works
+# NOTE: this check is not reliable in virt environment
+# since interfaces are always LOWER_UP. There is no way
+# from the guest to know if the host has disconnected somehow
+
+iface_lower_is_up() {
+	if ! $IP2UTIL -o link show $1 | \
+	    grep -q LOWER_UP; then
+		return 1
+	fi
+	return 0
+}
+
+macvlan_validate() {
+	check_binary $IP2UTIL
+
+	if [ -z "$OCF_RESKEY_device" ]; then
+		ocf_log err "Invalid OCF_RESKEY_device: value cannot be empty"
+		return 1
+	fi
+
+	if [ -z "$OCF_RESKEY_name" ]; then
+		ocf_log err "Invalid OCF_RESKEY_name: value cannot be empty"
+		return 1
+	fi
+
+	# the echo .. is the equivalent of strlen in bash
+	#
+	# /usr/include/linux/if.h:#define IFNAMSIZ        16
+	# needs to include 0 byte end string
+
+	if [ "${#OCF_RESKEY_name}" -gt 15 ]; then
+		ocf_log err "Invalid OCF_RESKEY_name: name is too long"
+		return 1
+	fi
+
+	if [ ! -d "/sys/class/net" ]; then
+		ocf_log err "Unable to find sysfs network class in /sys"
+		return 1
+	fi
+
+	if [ ! -e "/sys/class/net/$OCF_RESKEY_device" ]; then
+		ocf_log err "Invalid OCF_RESKEY_device: $OCF_RESKEY_device does not exists"
+		return 1
+	fi
+
+	if [ "${OCF_RESKEY_mode}" != "bridge" ]; then
+		ocf_log err "Invalid OCF_RESKEY_mode: only bridge mode is currently supported (because other modes are not tested)"
+		return 1
+	fi
+
+	return 0
+}
+
+macvlan_check() {
+	if [ -e "/sys/class/net/$OCF_RESKEY_name" ]; then
+		if [ ! -e "$HA_RSCTMP/iface-macvlan.$OCF_RESKEY_name" ]; then
+			return $OCF_ERR_GENERIC
```

Comment:
You might want to use OCF_ERR_CONFIGURED or _INSTALLED in some occurences in this function as well (depending on whether it would be caused by misconfiguration or lack of e.g. cli util).

Reply:
I only found one place where I have been sure to use another return code than OCF_ERR_GENERIC. I have already changed that.